merge upstream

frehner opened the pull request.

joeldenning commented on packages/react-router-redux/modules/ConnectedRouter.js:
```diff
@@ -26,7 +26,7 @@ class ConnectedRouter extends Component {
   };
 
-  componentWillMount() {
+  componentDidMount() {
```
This is one instance where I might consider switching to constructor instead of componentDidMount. The history of componentWillMount is that it was created before javascript classes were a thing and so cWM **was** the constructor. Now that we've had classes, cWM and constructor are sort of duplicates of each other, which is my understanding of why cWM got deprecated.

The difference from a code level is that if it's in cDM instead of CWM, `handleLocationChange()` (which dispatches a redux action) will be called only if the initial render is successful and only after everything is already rendered. So it might cause a situation where components who expect the location props to be correct on the first render to be wrong on the first render and later be corrected on subsequent renders. Which might be a problem.

That being said -- the maintainers of react-router will know better if this has a negative impact or not. I'm doing a deep dive here because i think the maintainers to react-router will also do a deep dive and will pay attention to detail for it. But they might not notice or care about this minor change in behavior

frehner replied:
Thanks! I'll look into moving this to constructor

joeldenning commented on packages/react-router/modules/Prompt.js:
```diff
@@ -46,10 +46,10 @@ class Prompt extends React.Component {
     if (this.props.when) this.enable(this.props.message);
```
this is another situation where that nuanced difference between cWM/constructor/cDM potentially could matter, but probably doesn't

joeldenning commented on packages/react-router/modules/Prompt.js:
```diff
@@ -46,10 +46,10 @@ class Prompt extends React.Component {
+  componentDidUpdate(prevProps) {
+    if (this.props.when) {
+      if (!prevProps.when || prevProps.message !== this.props.message)
+        this.enable(this.props.message);
```
I wasn't able to track down if `this.enable()` eventually calls set state or forces a re-render via React context. [This line](https://github.com/frehner/react-router/blob/af39f872275c788d9bf9ee3753d1d265b1c77829/packages/react-router/modules/Prompt.js#L30) is where I stopped in my search for that.

Why that would matter is that calling setState in cWRP is advantageous over calling it in cDM because it will prevent **another** render. In other words, if you call setState in cWRP it was already going to re-render so it doesn't trigger yet another render after that one. But if you call setState in cDU, it definitely will cause another render. Which is bad for performance. Normally, that nuanced difference doesn't matter a ton, but in the case of an extremely highly reused component inside of react-router, it might be a thing worth thinking about.

frehner replied:
Yeah, I'm just not entirely sure where else this _could_ go that would still work. gDSFP is static so I can't access `this` to call `this.enable`, so... 🤷‍♀️

joeldenning commented on packages/react-router/modules/Redirect.js:
```diff
@@ -31,20 +31,23 @@ class Redirect extends React.Component {
+  constructor(props, context) {
+    super(props, context);
+
+    if (this.isStatic()) this.perform();
```
:+1: this switch to using a constructor instead of cDM makes sense to me -- it was in CWM which is the equivalent to constructor

joeldenning commented on packages/react-router/modules/Route.js:
```diff
@@ -92,20 +93,24 @@ class Route extends React.Component {
+    const newMatch = this.computeMatch(this.props, this.context.router);
+
+    if (!isEqual(newMatch, this.state.match)) {
+      this.setState({
```
Calling setState inside of cDU is worse for performance than doing so in cWRP (see explanation above). Not sure I have an alternative (besides the rather garbage getDerivedStateFromProps lifecycle), but maybe something to think about

frehner replied:
yeah, I did think about using gDSFP in this case, but I do need access to `this.context` and `this.props` and I didn't think that putting a copy of those into state would be helpful... so it stayed here.

joeldenning commented on packages/react-router/modules/Router.js:
```diff
@@ -20,6 +20,26 @@ class Router extends React.Component {
+    // Do this here so we can setState when a <Redirect> changes the
+    // location in componentDidMount. This happens e.g. when doing
+    // server rendering using a <StaticRouter>.
+    this.unlisten = history.listen(() => {
```
:+1: I like this move to constructor instead of cDM

joeldenning commented on packages/react-router/modules/Router.js:
```diff
@@ -46,27 +66,9 @@ class Router extends React.Component {
-  }
-
-  componentWillReceiveProps(nextProps) {
+  componentDidUpdate(prevProps) {
```
:+1: this cDU is great because we're not calling setState in it. Very acceptable change for a reviewer

joeldenning commented on packages/react-router/modules/Switch.js:
```diff
@@ -19,21 +19,20 @@ class Switch extends React.Component {
   }
 
-  componentWillReceiveProps(nextProps) {
+  componentDidUpdate(prevProps) {
```
same with this one -- a very easy change for a reviewer to swallow because it doesn't impact performance

joeldenning commented on website/modules/animated/createAnimatedComponent.js:
```diff
@@ -35,7 +35,7 @@ function createAnimatedComponent(Component: any): any {
 
-    componentWillMount() {
+    componentDidMount() {
       this.attachProps(this.props);
```
This function does some [weird things](https://github.com/frehner/react-router/blob/af39f872275c788d9bf9ee3753d1d265b1c77829/website/modules/animated/createAnimatedComponent.js#L42), which would maybe require a performance deep dive to verify the lifecycle changes are okay. But since this is in react-router/website, I think it's fine because this isn't code that gets reused all over the world

joeldenning commented on website/modules/components/EnvironmentSmall.js:
```diff
@@ -46,16 +46,16 @@ class EnvironmentSmall extends Component {
   }
 
-  componentWillReceiveProps(nextProps) {
+  componentDidUpdate(prevProps) {
```
I'm skipping these changes because it's just for the website which doesn't matter as much.

Might be good to boot up their website locally though to verify you haven't broken things

frehner replied:
bad news: the links on the homepage don't work. good news: they don't work on master either. 🤷‍♀️

alireza-mh committed:
Refactor ...props to ...rest (#6421)

justsml committed:
Improve Initial Docs Experience (#6410)

* Added simpler "basic" example

* Trying to make Quick Start the default contents

"Philosophy" page is a bit too 'inside baseball' for an effective landing page.

* Moved Philosophy down further

* Added labels to 3rd party snippets

* Minor formatting fix

* fixing up quick start content

* updating "quick start" content

* a bit OCD formatting/alignment

frehner committed:
Added react-is to validate the proptype for Route's component. (#6417)

* Added react-is to validate the proptype for Route's component.

Removed a couple of outdated TODOs that dealt with __DEV__

The rest of the changes were done just from an npm install command

React-is does add a tiny bit of size ( https://bundlephobia.com/result?p=react-is@16.5.2 ) as noted by the snapshots, but it seems like the best (and official) way to support the check.

* Revert removal of __DEV__

* Updated the error message to be a little more clear about which prop is invalid.

* Realized it wasn't saying what I wanted, so I just reverted it.

rickhanlonii committed:
Upgrade Jest and switch to jest-cirucs (#6427)

mjackson committed:
Use version ranges instead of fixed versions

Commit:
Update size snapshot from fresh build

mjackson committed:
Remove unused `watch` script and unused deps

mjackson committed:
Use main `history` export

ElForastero committed:
Document generatePath function (#6437)

* Document generatePath function

* Simplify some things

alireza-mh committed:
Remove unused imports (#6436)

mjackson committed:
Add browserslist config

mjackson committed:
Bump history dep to 4.8.0-beta.0

Also, smaller UMD builds.

mjackson committed:
Whitespace

mjackson committed:
v4.4.0-beta.5

justsml committed:
Removed the Anywhere/Core button from Home Page (#6432)

alireza-mh committed:
Remove unused import (example directory) (#6441)

pshrmn committed:
Run tests against source, CJS, and UMD builds (#6209)

* Test react-router(-dom) source, CJS, and UMD

* Please go away, package-lock

* Test against new CJS builds

* Multi-test react-router-config

* Remove "resolved by" comments

* Retain original import order

mjackson committed:
Update package-lock and size-snapsnot

TrySound committed:
Replace dev expressions (#6444)

* Update lockfiles and size snapshot

* Replace invariant and warning with tiny versions

* Replace invariant and warning with dev versions in react-router-dom

mjackson committed:
Fix react-is import statement

Fixes #6445
Fixes #6446

mjackson committed:
Use build matrix to run tests in different envs

mjackson committed:
Add missing prop-types dep

Fixes #6438

mjackson committed:
v4.4.0-beta.6
